README.md
# Mandrill Mailer
[![Build Status](https://img.shields.io/github/workflow/status/renz45/mandrill_mailer/build.svg)](https://github.com/renz45/mandrill_mailer/actions/workflows/tests.yml) [![Gem Version](http://img.shields.io/gem/v/mandrill_mailer.svg)](https://rubygems.org/gems/mandrill_mailer)
[![Inline Documentation](http://inch-ci.org/github/renz45/mandrill_mailer.svg?branch=master)](http://inch-ci.org/github/renz45/mandrill_mailer)

Inherit the MandrillMailer class in your existing Rails mailers to send transactional emails through Mandrill using their template-based emails.

## Installation
Add this line to your application's Gemfile:

```
gem 'mandrill_mailer'
```

And then execute:

```
$ bundle install
```

Or install it yourself as:

```
$ gem install mandrill_mailer
```

## Usage
Add the following to your `mail.rb` in your Rails app's `config/initializers` directory:

```ruby
ActionMailer::Base.smtp_settings = {
    :address   => "smtp.mandrillapp.com",
    :port      => 587,
    :user_name => ENV['MANDRILL_USERNAME'],
    :password  => ENV['MANDRILL_API_KEY'],
    :domain    => 'heroku.com'
  }
ActionMailer::Base.delivery_method = :smtp

MandrillMailer.configure do |config|
  config.api_key = ENV['MANDRILL_API_KEY']
  config.deliver_later_queue_name = :default
end
```

You don't need to add the ActionMailer stuff unless you're still using ActionMailer emails.

This uses the Mandrill SMTP servers. If you're using template-based emails
through the Mandrill API you only need the `MandrillMailer.configure` portion.

Do not forget to setup the environment (`ENV`) variables on your server instead
of hardcoding your Mandrill username and password in the `mail.rb` initializer.

You will also need to set `default_url_options` for the mailer, similar to ActionMailer
in your environment config files in `config/environments`:

```ruby
config.mandrill_mailer.default_url_options = { :host => 'localhost' }
```

## Creating a new mailer
Creating a new Mandrill mailer is similar to a typical Rails one:

```ruby
class InvitationMailer < MandrillMailer::TemplateMailer
  default from: 'support@example.com'

  def invite(invitation)
    # in this example `invitation.invitees` is an Array
    invitees = invitation.invitees.map { |invitee| { email: invitee.email, name: invitee.name } }

    mandrill_mail(
      template: 'group-invite',
      subject: I18n.t('invitation_mailer.invite.subject'),
      to: invitees,
        # to: invitation.email,
        # to: { email: invitation.email, name: 'Honored Guest' },
      vars: {
        'OWNER_NAME' => invitation.owner_name,
        'PROJECT_NAME' => invitation.project_name
      },
      important: true,
      inline_css: true,
      recipient_vars: invitation.invitees.map do |invitee|
        { invitee.email =>
          {
            'INVITEE_NAME' => invitee.name,
            'INVITATION_URL' => new_invitation_url(
              invitee.email,
              secret: invitee.secret_code
            )
          }
        }
      end
     )
  end
end
```

* `#default:`
  * `:from` - set the default from email address for the mailer.  Defaults to `'example@email.com'`.
  * `:from_name` - set the default from name for the mailer. If not set, defaults to from email address. Setting :from_name in the .mandrill_mail overrides the default.
  * `:merge_vars` - set some default `merge_vars` that will be sent with every mailer method (in `global_merge_vars` so there's no risk of collision with method-specific `merge_vars`.
  * `:view_content_link` - set a default to be able to access individual mailer messages in the Mandrill dashboard. Defaults to `false`.

* `.mandrill_mail`
   * `:template`(required) - Template slug from within Mandrill (for backwards-compatibility, the template name may also be used but the immutable slug is preferred)

   * `:subject` - Subject of the email. If no subject supplied, it will fall back to the template default subject from within Mandrill

   * `:to`(required) - Accepts an email String, a Hash with :name and :email keys, or an Array of Hashes with :name, :email, and :type keys
      - examples:
        1. `'example@domain.com'`
        2. `{ email: 'someone@email.com', name: 'Bob Bertly' }`
        3. `[{ email: 'someone@email.com', name: 'Bob Bertly' }, { email: 'other@email.com', name: 'Claire Nayo' }]`
        4. `[{ email: 'someone@email.com', name: 'Bob Bertly', type: 'to'}, { email: 'secret_recipient1@email.com', name: 'Secret Recipient One', type: 'bcc'}, { email: 'secret_recipient2@email.com', name: 'Secret Recipient Two', type: 'bcc'}]`

   * `:vars` - A Hash of merge tags made available to the email. Use them in the
     email by wrapping them in `*||*`. For example `{'OWNER_NAME' => 'Suzy'}` is used by doing: `*|OWNER_NAME|*` in the email template within Mandrill

   * `:recipient_vars` - Similar to `:vars`, this is a Hash of merge vars specific to a particular recipient.
     Use this if you are sending batch transactions and hence need to send multiple emails at one go.
     ex. `[{'someone@email.com' => {'INVITEE_NAME' => 'Roger'}}, {'another@email.com' => {'INVITEE_NAME' => 'Tommy'}}]`

   * `:template_content` - A Hash of values and content for Mandrill editable content blocks.
     In MailChimp templates there are editable regions with 'mc:edit' attributes that look
     like: `<div mc:edit="header">My email content</div>` You can insert content directly into
     these fields by passing a Hash `{'header' => 'my email content'}`

   * `:headers` - Extra headers to add to the message (currently only `Reply-To` and `X-*` headers are allowed) {"...": "..."}

   * `:bcc` - Add an email to bcc to

   * `:tags` - Array of Strings to tag the message with. Stats are
   accumulated using tags, though we only store the first 100 we see,
   so this should not be unique or change frequently. Tags should be
   50 characters or less. Any tags starting with an underscore are
   reserved for internal use and will cause errors.

   * `:google_analytics_domains` - Array of Strings indicating for which any
   matching URLs will automatically have Google Analytics parameters appended
   to their query string automatically.

   * `:google_analytics_campaign` - String indicating the value to set for
   the utm_campaign tracking parameter. If this isn't provided the email's
   from address will be used instead.

   * `:important` - whether or not this message is important, and should be delivered ahead of non-important messages.

   * `:inline_css` - whether or not to automatically inline all CSS styles provided in the message HTML - only for HTML documents less than 256KB in size.

   * `:merge_language` - the merge tag language to use when evaluating merge tags, either 'mailchimp' or 'handlebars'. Default is 'mailchimp'.

   * `:attachments` - An array of file objects with the following keys:
     * `content`: The file contents, this will be encoded into a base64 string internally
     * `encoded_content` (optional): File content already encoded into base64 string. Overrides `content`
     * `name`: The name of the file
     * `type`: This is the mimetype of the file. Ex. png = image/png, pdf = application/pdf, txt = text/plain etc etc

   * `:images` - An array of embedded images to add to the message:
     * `content`: The file contents, this will be encoded into a base64 string internally
     * `encoded_content` (optional): File content already encoded into base64 string. Overrides `content`
     * `name`: The name of the file
     * `type`: This is the mimetype of the file. Ex. png = image/png, pdf = application/pdf, txt = text/plain etc etc etc

   * `:async` - Whether or not this message should be sent asynchronously

   * `:ip_pool` - The name of the dedicated ip pool that should be used to send the message

   * `:send_at` - When this message should be sent

## Sending a message without template
Sending a message without template is similar to sending a one with a template. The biggest
change is that you have to inherit from `MandrillMailer::MessageMailer` instead of the
MandrillMailer::TemplateMailer class:

```ruby
class InvitationMailer < MandrillMailer::MessageMailer
  default from: 'support@example.com'

  def invite(invitation)
    # in this example `invitation.invitees` is an Array
    invitees = invitation.invitees.map { |invitee| { email: invitee.email, name: invitee.name } }

    # no need to set up template and template_content attributes, set up the html and text directly
    mandrill_mail subject: I18n.t('invitation_mailer.invite.subject'),
                  to: invitees,
                  # to: invitation.email,
                  # to: { email: invitation.email, name: 'Honored Guest' },
                  text: "Example text content",
                  html: "<p>Example HTML content</p>",
                  # when you need to see the content of individual emails sent to users
                  view_content_link: true,
                  vars: {
                    'OWNER_NAME' => invitation.owner_name,
                    'PROJECT_NAME' => invitation.project_name
                  },
                  important: true,
                  inline_css: true,
                  attachments: [
                    {
                      content: File.read(File.expand_path('assets/offer.pdf')),
                      name: 'offer.pdf',
                      type: 'application/pdf'
                    }
                  ],
                  recipient_vars: invitation.invitees.map do |invitee| # invitation.invitees is an Array
                    { invitee.email =>
                      {
                        'INVITEE_NAME' => invitee.name,
                        'INVITATION_URL' => new_invitation_url(invitee.email, secret: invitee.secret_code)
                      }
                    }
                  end
  end
end
```

## Sending an email

You can send the email by using the familiar syntax:

`InvitationMailer.invite(invitation).deliver_now`
`InvitationMailer.invite(invitation).deliver_later(wait: 1.hour)`
For deliver_later, Active Job will need to be configured

## Creating a test method
When switching over to Mandrill for transactional emails we found that it was hard to setup a mailer in the console to send test emails easily (those darn designers), but really, you don't want to have to setup test objects everytime you want to send a test email. You can set up a testing 'mock' once and then call the `.test` method to send the test email.

You can test the above email by typing: `InvitationMailer.test(:invite, email:<your email>)` into the Rails Console.

The test for this particular Mailer is setup like so:

```ruby
test_setup_for :invite do |mailer, options|
    invitation = MandrillMailer::Mock.new({
      email: options[:email],
      owner_name: 'foobar',
      secret: rand(9000000..1000000).to_s
    })
    mailer.invite(invitation).deliver
end
```

Use MandrillMailer::Mock to mock out objects.

If in order to represent a url within a mock, make sure there is a `url` or `path` attribute,
for example, if I had a course mock and I was using the `course_url` route helper within the mailer
I would create the mock like so:

```ruby
course = MandrillMailer::Mock.new({
  title: 'zombies',
  type: 'Ruby',
  url: 'http://funzone.com/zombies'
})
```

This would ensure that `course_url(course)` works as expected.

The mailer and options passed to the `.test` method are yielded to the block.

The `:email` option is the only required option, make sure to add at least this to your test object.

## Offline Testing
You can turn on offline testing by requiring this file (say, in your spec_helper.rb):

```ruby
require 'mandrill_mailer/offline'
```

And then if you wish you can look at the contents of `MandrillMailer.deliveries` to see whether an email was queued up by your test:

```ruby
email = MandrillMailer.deliveries.detect { |mail|
  mail.template_name == 'my-template' &&
  mail.message['to'].any? { |to| to[:email] == 'my@email.com' }
}
expect(email).to_not be_nil
```

Don't forget to clear out deliveries:

```ruby
before :each { MandrillMailer.deliveries.clear }
```

## Using Delayed Job
The typical Delayed Job mailer syntax won't work with this as of now. Either create a custom job or queue the mailer as you would queue a method. Take a look at the following examples:

```ruby
def send_hallpass_expired_mailer
  HallpassMailer.hallpass_expired(user).deliver
end
handle_asynchronously :send_hallpass_expired_mailer
```

or using a custom job

```ruby
def update_email_on_newsletter_subscription(user)
  Delayed::Job.enqueue( UpdateEmailJob.new(user_id: user.id) )
end
```
The job looks like (Don't send full objects into jobs, send ids and requery inside the job. This prevents Delayed Job from having to serialize and deserialize whole ActiveRecord Objects and this way, your data is current when the job runs):

```ruby
class UpdateEmailJob < Struct.new(:user_id)
  def perform
    user = User.find(user_id)
    HallpassMailer.hallpass_expired(user).deliver
  end
end
```

## Using Sidekiq
Create a custom worker:

```ruby
class UpdateEmailJob
  include Sidekiq::Worker
  def perform(user_id)
    user = User.find(user_id)
    HallpassMailer.hallpass_expired(user).deliver
  end
end

#called by
UpdateEmailJob.perform_async(<user_id>)
```

Or depending on how up to date things are, try adding the following to `config/initializers/mandrill_mailer_sidekiq.rb`

```ruby
::MandrillMailer::TemplateMailer.extend(Sidekiq::Extensions::ActionMailer)
```

This should enable you to use this mailer the same way you use ActionMailer.
More info: https://github.com/mperham/sidekiq/wiki/Delayed-Extensions#actionmailer

## Using Resque

Create a job:
```ruby
class SendUserMailJob
  def initialize(user_id)
    @user_id = user_id
  end

  def work
    user = User.find(@user_id)
    UserMailer.send_user_email(user).deliver
  end
end
```

Send your job to Resque:
```ruby
resque = Resque.new
resque << SendUserMailJob.new(<user id>)
```

## Using an interceptor
You can set a mailer interceptor to override any params used when you deliver an e-mail.
The interceptor is a Proc object that gets called with the mail object being sent
to the api.

Example that adds multiple bcc recipients:

```ruby
MandrillMailer.configure do |config|
  config.interceptor = Proc.new {|params|

    params["to"] =  [
      params["to"],
      { "email" => "bccEmailThatWillBeUsedInAll@emailsSent1.com", "name" => "name", "type" => "bcc" },
      { "email" => "bccEmailThatWillBeUsedInAll@emailsSent2.com", "name" => "name", "type" => "bcc" },
      { "email" => "bccEmailThatWillBeUsedInAll@emailsSent3.com", "name" => "name", "type" => "bcc" }
    ].flatten
  }
end
```


CHANGELOG.md
# Change Log
All notable changes to this project will be documented in this file.
This project adheres to [Semantic Versioning](http://semver.org/).

# Unreleased

## 1.8.0 - 2021-03-18
- Support Ruby 3.0.0 [#142](https://github.com/renz45/mandrill_mailer/pull/142) (via @berkos)
- Allow symbols in vars [#140](https://github.com/renz45/mandrill_mailer/pull/140) (via @raphaelpra)

## 1.7.1 - 2018-10-06
- Fix rspec matcher: https://github.com/renz45/mandrill_mailer/pull/136

## 1.7.0 - 2018-10-06
- Switch to mandrillus fork of the mandrill-api gem to allow for newer
json gem support

## 1.6.0 - 2017-03-23
- Add support for attaching unencoded content via @brunomperes

## 1.5.0 - 2017-03-20
- Update offline adapter for testing to be more compatible with Rails 5 via @eric1234

## 1.4.0 - 2016-04-28
### Changed
- Update deprecated RSpec failure methods in RSpec helpers.

## 1.3.0 - 2016-03-02
### Added
- Fixed an issue where deliver_later functionality was not working as intended when inheriting from the mailer classes. via @eric1234

## 1.2.0 - 2015-12-22
### Added
- Support for for deliver_later and deliver_now using ActiveJob. via @BenRuns


## 1.1.0 - 2015-10-02
### Added
- Optional RSpec helper (`MandrillMailer::RSpecHelper`) with custom matchers
to simplify testing mailers like:
  - `expect(mailer).to be_from("email@example.com")`
  - `expect(mailer).to have_merge_data('USER_EMAIL' => user.email)`
  - `expect(mailer).to include_merge_var_content(user.email)`
  - `expect(mailer).to have_subject("Hello")`
  - `expect(mailer).to use_template('Example')`
  - `expect(mailer).to send_email_to('email@example.com')`


## 1.0.4 - 2015-09-24
### Added
- Allow default `view_content_link` on Mailer class.

## 1.0.3
- Fix a bug where defaults in merge vars were receiving the correct defaults (credit @kennethkalmer).

## 1.0.1
- Correct regression caused in 1.0.0 that broke defaults in mailers (credit: @etipton).

## 1.0.0
### Changed
- Update manrill_api gem to 1.0.X.
- Change how interceptors work to be more flexible and not overwrite data if needed.
- Make both the template and message mailers compatible with all available attributes in the messages API.

### Removed
- Deprecated `data` method on Mailer objects, replaced with `message`.

## 0.6.1
### Fixed
- Correct a regression introduced in 0.6.0 that caused a TypeError exception.
when no `merge_vars` were provided as arguments to `mandrill_mail`.

## 0.6.0
### Changed
- Allow for default `merge_vars` to be set on the mailer class so that every method inherits them, just like the existing default `from` and `from_name`.

## 0.5.2
### Changed
- Allow the mandrill api gem to be more flexible in the accepted version.
  which allows for the json gem to be updated internally.

## 0.5.0
### Changed
- Made the file attachment and image apis compatible with mandrill api doc syntax.

## 0.4.9
### Added
- Offline support for the `MandrillMailer::Messenger` class. Thanks @arthurtalkgoal.
- Support for the Mandrill Messages api by adding a `MandrillMailer::Messenger` class. Thanks @arthurtalkgoal.

## 0.4.8
### Added
- `to=` setter to template mailer.

## 0.4.7
### Changed
- Reworked how defaults were stored so that they are accessible when being extended.

## 0.4.6
### Added
- Support for images array of embedded images.

## 0.4.5
### Fixed
- Declare `mandrill-api` gem as a runtime dependency to prevent load errors.

## 0.4.4
### Added
- Added `to`, `from` and `bcc` getters to `TemplateMailer` (@renz45).
- Documented how to use the gem with Sidekiq (@elado, @renz45).
- Documented fallback to Mandrill default subject on `mandrill_mail` subject (@Advocation).

### Fixed
- Fix `respond_to` method as it doesn't follow the way standard Ruby respond_to behaves (@tpaktop).

## 0.4.3
### Added
- Look at deliveries for offline testing.
- Intercept emails before they are sent in the config (see the specs for more info).

## 0.4.1
### Added
- Subaccount parameter.

## 0.4.0
### Changed
- Default setting for preserve_recipients no longer defaults to true. Bumping a minor version since this might have been assumed to be default and not set by some.

## 0.3.8
### Added
- This change log.


[unreleased]: https://github.com/olivierlacan/keep-a-changelog/compare/v0.6.0...HEAD


lib/mandrill_mailer/core_mailer.rb
# MandrilMailer class for sending transactional emails through mandril.
# Only template based emails are supported at this time.

# Example usage:

# class InvitationMailer < MandrillMailer::TemplateMailer
#   default from: 'support@codeschool.com',
#           from_name: 'Code School',
#           merge_vars: { 'FOO' => 'Bar' }
#           view_content_link: false

#   def invite(invitation)
#     invitees = invitation.invitees.map { |invitee| { email: invitee.email, name: invitee.name } }
#
#     mandrill_mail template: 'Group Invite',
#                   subject: I18n.t('invitation_mailer.invite.subject'),
#                   to: invitees,
#                   # to: invitation.email
#                   # to: { email: invitation.email, name: invitation.recipient_name }
#                   vars: {
#                     'OWNER_NAME' => invitation.owner_name,
#                     'PROJECT_NAME' => invitation.project_name
#                   },
#                   recipient_vars: invitation.invitees.map do |invitee| # invitation.invitees is an Array
#                                     { invitee.email =>
#                                       {
#                                         'INVITEE_NAME' => invitee.name,
#                                         'INVITATION_URL' => new_invitation_url(invitee.email, secret: invitee.secret_code)
#                                       }
#                                     }
#                                   end,
#                   template_content: {},
#                   attachments: [{file: File.read(File.expand_path('assets/some_image.png')), filename: 'My Image.png', mimetype: 'image/png'}],
#                   important: true,
#                   inline_css: true
#   end
# end

# #default:
#   :from               - set the default from email address for the mailer
#   :from_name          - set the default from name for the mailer
#   :merge_vars         - set the default merge vars for the mailer
#   :view_content_link  - set a default view_content_link option for the mailer

# .mandrill_mail
#   :template(required) - Template name from within Mandrill

#   :subject(required) - Subject of the email

#   :to(required) - Accepts an email String, a Hash with :name and :email keys
#                   or an Array of Hashes with :name and :email keys
#     examples:
#       1)
#         'example@domain.com`
#       2)
#         { email: 'someone@email.com', name: 'Bob Bertly' }
#       3)
#         [{ email: 'someone@email.com', name: 'Bob Bertly' },
#          { email: 'other@email.com', name: 'Claire Nayo' }]
#

#   :vars - A Hash of merge tags made available to the email. Use them in the
#     email by wrapping them in '*||*' vars: {'OWNER_NAME' => 'Suzy'} is used
#     by doing: *|OWNER_NAME|* in the email template within Mandrill
#
#   :recipient_vars - Similar to :vars, this is a Hash of merge tags specific to a particular recipient.
#     Use this if you are sending batch transactions and hence need to send multiple emails at one go.
#     ex. [{'someone@email.com' => {'INVITEE_NAME' => 'Roger'}}, {'another@email.com' => {'INVITEE_NAME' => 'Tommy'}}]

#   :template_content - A Hash of values and content for Mandrill editable content blocks.
#     In MailChimp templates there are editable regions with 'mc:edit' attributes that look
#     a little like: '<div mc:edit="header">My email content</div>' You can insert content directly into
#     these fields by passing a Hash {'header' => 'my email content'}

#   :attachments - An array of file objects with the following keys:
#       file: This is the actual file, it will be converted to byte data in the mailer
#       filename: The name of the file
#       mimetype: This is the mimetype of the file. Ex. png = image/png, pdf = application/pdf, txt = text/plain etc

#   :images - An array of embedded images to add to the message:
#       file: This is the actual file, it will be converted to byte data in the mailer
#       filename: The Content ID of the image - use <img src="cid:THIS_VALUE"> to reference the image in your HTML content
#       mimetype: The MIME type of the image - must start with "image/"

# :headers - Extra headers to add to the message (currently only Reply-To and X-* headers are allowed) {"...": "..."}

# :bcc - Add an email to bcc to

# :tags - Array of Strings to tag the message with. Stats are
#   accumulated using tags, though we only store the first 100 we see,
#   so this should not be unique or change frequently. Tags should be
#   50 characters or less. Any tags starting with an underscore are
#   reserved for internal use and will cause errors.

# :google_analytics_domains - Array of Strings indicating for which any
#   matching URLs will automatically have Google Analytics parameters appended
#   to their query string automatically.

# :google_analytics_campaign - String indicating the value to set for
#   the utm_campaign tracking parameter. If this isn't provided the email's
#   from address will be used instead.

# :inline_css - whether or not to automatically inline all CSS styles provided in the
#   message HTML - only for HTML documents less than 256KB in size

# :important - whether or not this message is important, and should be delivered ahead of non-important messages

# Required for hash.stringify_keys!
require 'active_support/all'
require 'mandrill_mailer/arg_formatter'

module MandrillMailer
  class CoreMailer
    class InvalidEmail < StandardError; end
    class InvalidMailerMethod < StandardError; end
    class InvalidInterceptorParams < StandardError; end
    class InvalidMergeLanguageError < StandardError; end

    # Public: Other information on the message to send
    attr_accessor :message

    # Public: Enable background sending mode
    attr_accessor :async

    # Public:  Name of the dedicated IP pool that should be used to send the message
    attr_accessor :ip_pool

    # Public: When message should be sent
    attr_accessor :send_at

    # Public: Defaults for the mailer. Currently the only option is from:
    #
    # options       - The Hash options used to refine the selection (default: {}):
    #   :from       - Default from email address
    #   :from_name  - Default from name
    #   :merge_vars - Default merge vars
    #
    # Examples
    #
    #   default from: 'foo@bar.com',
    #           from_name: 'Foo Bar',
    #           merge_vars: {'FOO' => 'Bar'}
    #
    #
    # Returns options
    def self.defaults
      @defaults || super_defaults || {}
    end

    def self.super_defaults
      superclass.defaults if superclass.respond_to?(:defaults)
    end

    def self.default(args)
      @defaults ||= {}
      @defaults[:from] ||= 'example@email.com'
      @defaults[:merge_vars] ||= {}
      @defaults.merge!(args)
    end

    class << self
      attr_writer :defaults
    end

    # Public: setup a way to test mailer methods
    #
    # mailer_method - Name of the mailer method the test setup is for
    #
    # block - Block of code to execute to perform the test. The mailer
    # and options are passed to the block. The options have to
    # contain at least the :email to send the test to.
    #
    # Examples
    #
    #   test_setup_for :invite do |mailer, options|
    #     invitation = OpenStruct.new({
    #       email: options[:email],
    #       owner_name: 'foobar',
    #       secret: rand(9000000..1000000).to_s
    #     })
    #     mailer.invite(invitation).deliver
    #   end
    #
    # Returns the duplicated String.
    def self.test_setup_for(mailer_method, &block)
      @mailer_methods ||= {}
      @mailer_methods[mailer_method] = block
    end

    # Public: Executes a test email
    #
    # mailer_method - Method to execute
    #
    # options - The Hash options used to refine the selection (default: {}):
    #   :email - The email to send the test to.
    #
    # Examples
    #
    #   InvitationMailer.test(:invite, email: 'benny@envylabs.com')
    #
    # Returns the duplicated String.
    def self.test(mailer_method, options={})
      unless options[:email]
        raise InvalidEmail.new 'Please specify a :email option(email to send the test to)'
      end

      if @mailer_methods[mailer_method]
        @mailer_methods[mailer_method].call(self.new, options)
      else
        raise InvalidMailerMethod.new "The mailer method: #{mailer_method} does not have test setup"
      end

    end

    # Public: Triggers the stored Mandrill params to be sent to the Mandrill api
    def deliver
      raise NotImplementedError.new("#{self.class.name}#deliver is not implemented.")
    end
    
    def deliver_now
      raise NotImplementedError.new("#{self.class.name}#deliver_now is not implemented.")
    end
    
    def deliver_later
      raise NotImplementedError.new("#{self.class.name}#deliver_later is not implemented.")
    end

    def mandrill_mail_handler(args)
      args
    end

    # Public: Build the hash needed to send to the mandrill api
    #
    # args - The Hash options used to refine the selection:
    #
    # Examples
    #
    #   mandrill_mail template: 'Group Invite',
    #               subject: I18n.t('invitation_mailer.invite.subject'),
    #               to: invitation.email,
    #               vars: {
    #                 'OWNER_NAME' => invitation.owner_name,
    #                 'INVITATION_URL' => new_invitation_url(email: invitation.email, secret: invitation.secret)
    #               }
    #
    # Returns the the mandrill mailer class (this is so you can chain #deliver like a normal mailer)
    def mandrill_mail(args)
      extract_api_options!(args)

      # Call the mandrill_mail_handler so mailers can handle the args in custom ways
      mandrill_mail_handler(args)

      # Construct message hash
      self.message = MandrillMailer::ArgFormatter.format_messages_api_message_data(args, self.class.defaults)

      # Apply any interceptors that may be present
      apply_interceptors!(self.message)

      # return self so we can chain deliver after the method call, like a normal mailer.
      self
    end

    def from
      self.message && self.message['from_email']
    end

    def to
      self.message && self.message['to']
    end

    def to=(values)
      self.message && self.message['to'] = MandrillMailer::ArgFormatter.params(values)
    end

    def bcc
      self.message && self.message['bcc_address']
    end

    protected

    def apply_interceptors!(obj)
      unless MandrillMailer.config.interceptor.nil?
        unless MandrillMailer.config.interceptor.is_a?(Proc)
          raise InvalidInterceptorParams.new "The interceptor_params config must be a proc"
        end
        MandrillMailer.config.interceptor.call(obj)
      end

      obj
    end

    # Makes this class act as a singleton without it actually being a singleton
    # This keeps the syntax the same as the original mailers so we can swap quickly if something
    # goes wrong.

    class << self
      def method_missing(method, *args, &block)
        return super unless respond_to?(method)
        new.method(method).call(*args, &block)
      end

      ruby2_keywords(:method_missing) if respond_to?(:ruby2_keywords, true)
    end

    def self.respond_to?(method, include_private = false)
      super || instance_methods.include?(method.to_sym)
    end

    # Proxy route helpers to rails if Rails exists. Doing routes this way
    # makes it so this gem doesn't need to be a rails engine
    def method_missing(method, *args)
      return super unless defined?(Rails) && Rails.application.routes.url_helpers.respond_to?(method)
      # Check to see if one of the args is an open struct. If it is, we'll assume it's the
      # test stub and try to call a path or url attribute.
      if args.any? {|arg| arg.kind_of?(MandrillMailer::Mock)}
        # take the first OpenStruct found in args and look for .url or.path
        args.each do |arg|
          if arg.kind_of?(MandrillMailer::Mock)
            break arg.url || arg.path
          end
        end
      else
        options = args.extract_options!.merge({host: MandrillMailer.config.default_url_options[:host], protocol: MandrillMailer.config.default_url_options[:protocol]})
        args << options
        Rails.application.routes.url_helpers.method(method).call(*args)
      end
    end

    def image_path(image)
      if defined? Rails
        ActionController::Base.helpers.asset_path(image)
      else
        method_missing(:image_path, image)
      end
    end

    def image_url(image)
      "#{root_url}#{image_path(image).split('/').reject!(&:empty?).join('/')}"
    end

    def api_key
      MandrillMailer.config.api_key
    end

    def mandrill_api
      @mandrill_api ||= Mandrill::API.new(api_key)
    end

    def extract_api_options!(args)
      self.async = args.delete(:async)
      self.ip_pool = args.delete(:ip_pool)

      if args.has_key?(:send_at)
        self.send_at = args.delete(:send_at).getutc.strftime('%Y-%m-%d %H:%M:%S')
      end
    end
  end
end
